### ISSUE TITLE
Memory leak in Ipopt::Ma86SolverInterface

### ISSUE THREAD

**Issue by svigerske** [coin-or/Ipopt#283]
Issue created by migration from Trac.

Original creator: mhahn

Original creation time: 2017-06-28 14:27:25

Assignee: ipopt-team

Version: 3.12

It appears that the member array `order_` in the class `Ipopt::Ma86SolverInterface` is not being freed upon destruction of the solver interface. While not a major concern in single NLP solves, this appears to lead to a cumulative loss of free memory in applications where a large number of NLP solves must be performed (e.g. as relaxation solver in MINLP branch and bound).

To reproduce, build IPOPT (revision indicated in `svn-info`) after getting third party dependencies and configuring with the following arguments:
```
./configure --with-asl=BUILD --with-metis=BUILD --with-hsl=BUILD --with-mumps=BUILD --enable-debug
```
In this instance, BLAS and LAPACK are provided through the OS distribution (Ubuntu 17.04) and the HSL version is 2014.01.10.

Then, install IPOPT and dependencies to a prefix of your choice, and execute the following following command line to solve the problem described in the attached file `batch.nl`:
```
valgrind --tool=memcheck --leak-check=full ipopt batch.nl linear_solver=ma86 outlev=5
```
The expected output (mixed stdout/stderr) is attached in `valgrind.log`.

A possible fix is given in `possible-fix.diff`. It appears to get rid of the leak reported by `valgrind`. However, I have not verified whether there is a set of circumstances in which `order_ == NULL` when the destructor is invoked.

Attached are:
* `batch.nl`: A stub .nl file derived from the test problem 'batch' from the MacMINLP collection (https://wiki.mcs.anl.gov/leyffer/index.php/MacMINLP)
* `svn-info`: A dump of the output of `svn info` to indicate the branch and revision of IPOPT that was used
* `valgrind.log`: A dump of both stdout and stderr for the example run
* `possible-fix.diff`: A diff file indicating a possible fix that removes the leak reported in `valgrind.log`

**Comment by svigerske** [comment]
Attachment [batch.nl](https://gist.githubusercontent.com/svigerske/cca6fb6ad12c17897d8c04df00428342/raw/42e8cfcaf503c119008eeea5aadb1558c34c6e75/Ipopt%20issue%20283%20attachment%20batch.nl) by mhahn created at 2017-06-28 14:28:02

Test problem file

**Comment by svigerske** [comment]
Attachment [svn-info](https://gist.githubusercontent.com/svigerske/45d3e7330eb4ff1f1e4f7a586518c329/raw/5ff5c8257207bb077516e95432dd9d7f565f268b/Ipopt%20issue%20283%20attachment%20svn-info) by mhahn created at 2017-06-28 14:28:38

Info about SVN revision in question

**Comment by svigerske** [comment]
Attachment [valgrind.log](https://gist.githubusercontent.com/svigerske/4985285178b536630e5f35556f97f661/raw/cdfc0c6b9a70ca6184317ab08ec8661190d0cdd3/Ipopt%20issue%20283%20attachment%20valgrind.log) by mhahn created at 2017-06-28 14:29:02

Valgrind output

**Comment by svigerske** [comment]
Attachment [possible-fix.diff](https://gist.githubusercontent.com/svigerske/f5861da35e0ee610d18aef5edcd2c79c/raw/b583fb70b56860821faa02417312d2d4ed034f86/Ipopt%20issue%20283%20attachment%20possible-fix.diff) by mhahn created at 2017-06-28 14:29:53

Possible (but largely untested) fix

**Comment by svigerske** [comment]
Comment by @svigerske created at 2017-06-30 10:24:10

Thanks for finding this. This is now fixed in trunk 2c707bd4ea and should make it into the next release.

order_ can be NULL in the destructor, but a delete[] on a NULL pointer is ok (it just does nothing).

**Comment by svigerske** [comment]
Comment by @svigerske created at 2017-06-30 10:24:10

Resolution: fixed
